### PR TITLE
RPC: Error handling improvements

### LINKED ISSUE #1309: Attach another log handler to `log()` in RPC server context
When we compile or run SQL via the RPC server, any logs or warnings generated can be useful to the client. Those logs should be returned as a new key in the RPC response.

The following should *not* be part of the response:
 - Access logs + request / response logs

Everything else, just slap into the response

### LINKED ISSUE #1310: Add richer exception info to all of the dbt exceptions
Implement error codes according to the RPC spec. Pick a subrange of the available codes, assign each exception a code. Add a user readable, static message to each exception.

Make sure that dbt returns the right codes and messages

### PR DESCRIPTION
**Pull request by beckjake**
Fixes #1309 
Fixes #1310 

All dbt errors now have proper error codes/messages
The raised message at runtime ends up in result.error.data.message
The raised message type at runtime ends up in result.error.data.typename
result.error.message is a plaintext name for result.error.code
dbt.exceptions.Exception.data() becomes result.error.data
Include logs in the RPC responses - for errors under `response.error.data.logs` and for results under `response.result.logs`. The results are enqueued back to the main process, so even timeouts get logs.

I refactored safe_run by just chunking everything up as part of this, as I wanted to override particular parts of the error handling chain (mostly to keep raising errors farther and father up the stack of `safe_*`)

### CODE REVIEW

**Comment by beckjake** [comment]
>We should make logs a dict with three elements: debug, info, and manual
>The existing logs array should become logs['debug']
>We should siphon off info-level logs into the info array

Does this mean we should duplicate logs, or that users have to manually reconcile ordering between the `debug`/`info` levels if they care about that?



>We currently drop a backtrace into the debug logs which is good, but we'll want to tidy that up for info/manual level logs if possible

What do you mean by this? What exactly output do you want from an exception stack trace and where do you want to see it? Do you want them in the `debug` logs returned to the user, or not? Currently we don't to my knowledge log any stack traces at `info` level so I guess I'm confused as to why this is a concern at all in the context of your other comments.

>Instead of re-routing the info-level {{ logs() }} calls in the rpc server, we could alternatively make a print() context function for this purpose. I don't feel strongly about that either way! These functions should work sensibly in CLI or RPC contexts.

That sounds like it would be harder for no benefit? Currently I just added a log handler, so anything that would get logged goes there and gets sent over the queue. I assume `{{ print() }}` would have to know if it's in an RPC vs CLI context and react accordingly. Why bother?

>Can we nix the time string for info/manual logs?
Yeah. In `emit()` we get a `logRecord` and can do whatever we want with it. We don't even have to emit strings, we can emit any json-serializable object.

**Comment by cmcarthur** [comment]
> > Instead of re-routing the info-level {{ logs() }} calls in the rpc server, we could alternatively make a print() context function for this purpose. I don't feel strongly about that either way! These functions should work sensibly in CLI or RPC contexts.
> 
> That sounds like it would be harder for no benefit? Currently I just added a log handler, so anything that would get logged goes there and gets sent over the queue. I assume `{{ print() }}` would have to know if it's in an RPC vs CLI context and react accordingly. Why bother?

👍

@beckjake what does the typical INFO output from a `RemoteRunTask` look like right now?

File: core/dbt/task/rpc_server.py [hunk 1]
```diff
@@ -1,16 +1,126 @@
...
+        if timeout < 0:
+            raise dbt.exceptions.RPCTimeoutException(self.timeout)
+        return timeout
+
+    def _wait_for_results(self):
+        """Wait for results off the queue. If there is a timeout set, and it is
+        exceeded, raise an RPCTimeoutException.
+        """
+        while True:
+            get_timeout = self._next_timeout()
```
Comment (cmcarthur) [inline core/dbt/task/rpc_server.py]: the timeout solution here is great.

**Comment by beckjake** [comment]
@cmcarthur The only INFO line I see is: `2019-03-08 08:00:38,609 (MainThread): Found 3 models, 0 tests, 0 archives, 0 analyses, 95 macros, 0 operations, 1 seed files, 1 sources, 1 None'`

File: core/dbt/task/rpc_server.py [hunk 2]
```diff
@@ -1,16 +1,126 @@
...
+            raise rpc.timeout_error(self.timeout)
+        except dbt.exceptions.Exception as exc:
+            raise rpc.dbt_error(exc)
+        except Exception as exc:
+            raise rpc.server_error(exc)
+        finally:
+            self.process.join()
+
+        self.process = None
+        self.queue = None
```
Comment (cmcarthur) [inline core/dbt/task/rpc_server.py]: are these two lines significant from a GC perspective? if so it should probably go into the `finally` block

Comment (beckjake) [reply core/dbt/task/rpc_server.py]: Nope, they're not significant, I should just remove them.

**Pull request merged**